src/V3LinkInc.cpp

```diff
@@ -0,0 +1,220 @@
+            if (debug() >= 9) newp->dumpTree(cout, "-newfunc:");
+            m_insStmtp->addHereThisAsNext(newp);
+        } else if (m_insMode == IM_AFTER) {
+            m_insStmtp->addNextHere(newp);
+        } else if (m_insMode == IM_WHILE_PRECOND) {
+            AstWhile* whilep = VN_CAST(m_insStmtp, While);
+            UASSERT_OBJ(whilep, nodep, "Insert should be under WHILE");
+            whilep->addPrecondsp(newp);
+        } else {
+            nodep->v3fatalSrc("Unknown InsertMode");
+        }
+        m_insMode = IM_AFTER;
+        m_insStmtp = newp;
+    }
+
+    // VISITORS
+    virtual void visit(AstWhile* nodep) VL_OVERRIDE {
+        // Special, as statements need to be put in different places
+        // Preconditions insert first just before themselves (the normal
+        // rule for other statement types)
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->precondsp());
+        // Conditions insert first at end of precondsp.
+        m_insMode = IM_WHILE_PRECOND;
+        m_insStmtp = nodep;
+        iterateAndNextNull(nodep->condp());
+        // Body insert just before themselves
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->bodysp());
+        iterateAndNextNull(nodep->incsp());
+        // Done the loop
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    virtual void visit(AstNodeFor* nodep) VL_OVERRIDE {
+        nodep->v3fatalSrc(
+            "For statements should have been converted to while statements in V3Begin.cpp");
+    }
+    virtual void visit(AstNodeStmt* nodep) VL_OVERRIDE {
+        if (!nodep->isStatement()) {
+            iterateChildren(nodep);
+            return;
+        }
+        m_insMode = IM_BEFORE;
+        m_insStmtp = nodep;
+        iterateChildren(nodep);
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    void unsupported_visit(AstNode* nodep) {
+        m_unsupportedHere = true;
+        iterateChildren(nodep);
+        m_unsupportedHere = false;
+    }
+    virtual void visit(AstLogAnd* nodep) VL_OVERRIDE { unsupported_visit(nodep); }
+    void prepost_visit(AstNode* nodep) {
+        // Check if we are underneath a statement
+        if (m_insStmtp == NULL) {
+            prepost_non_stmt_visit(nodep);
+        } else {
+            prepost_stmt_visit(nodep);
+        }
+    }
+    void prepost_non_stmt_visit(AstNode* nodep) {
+        iterateChildren(nodep);
+
+        AstNRelinker linker;
+        AstNode* innerp = nodep->op1p();
+        AstConst* constp = VN_CAST(nodep->op2p(), Const);
+        V3Number num = constp->num();
+
+        // Make sure it can be lvalue if it's a VarRef
+        if (VN_IS(innerp, VarRef)) {
+            AstNodeVarRef* vr = VN_CAST(innerp, VarRef);
+            vr->lvalue(true);
+        }
+
+        num.width(32, false);
+        AstConst* newconstp = new AstConst(nodep->fileline(), num);
```

Comment:
```
        V3Number num = constp->num();
        num.width(32, false);
        AstConst* newconstp = new AstConst(nodep->fileline(), num);
```

You're resizing to 32 bits.  Try just

        AstConst* newconstp = constp->cloneTree();

Reply:
That was actually just to test whether I can fix this problem by manually setting the width. If I don't set I get errors for all widths:
```
%Warning-WIDTH: t/t_sv_conditional.v:269:20: Operator ADD expects 32 bits on the LHS, but LHS's VARREF 'cntA' generates 4 bits.
                                           : ... In instance t.st3_testbench_i.simple_test_3_i.counterA_inst
  269 |                cntA--;
      |                    ^~
                ... Use "/* verilator lint_off WIDTH */" and lint_on around source to disable this message.
%Warning-WIDTH: t/t_sv_conditional.v:269:20: Operator ASSIGN expects 4 bits on the Assign RHS, but Assign RHS's ADD generates 32 bits.
                                           : ... In instance t.st3_testbench_i.simple_test_3_i.counterA_inst
  269 |                cntA--;
      |                    ^~
%Warning-WIDTH: t/t_sv_conditional.v:276:20: Operator ADD expects 32 bits on the LHS, but LHS's VARREF 'cntA' generates 4 bits.
                                           : ... In instance t.st3_testbench_i.simple_test_3_i.counterA_inst
  276 |                cntA++;
      |                    ^~
%Warning-WIDTH: t/t_sv_conditional.v:276:20: Operator ASSIGN expects 4 bits on the Assign RHS, but Assign RHS's ADD generates 32 bits.
                                           : ... In instance t.st3_testbench_i.simple_test_3_i.counterA_inst
  276 |                cntA++;
      |                    ^~
%Error: Exiting due to 4 warning(s)
```

But it seems that if I tried to get the width of the actual variables I'm always getting a `1` - I tried with `vr->varp()->childDTypep()->width()` and `vr->varp()->declKwd().width()` (shooting in the dark to be honest) is that because the vars are not widthed yet?